File: internal/plugins/typescript/rules/strict_void_return/strict_void_return_test.go
package strict_void_return

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestStrictVoidReturnRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &StrictVoidReturnRule, []rule_tester.ValidTestCase{
		{Code: `declare function takesVoid(cb: () => void): void; takesVoid(() => {});`},
		{Code: `declare function takesVoid(cb: () => void): void; takesVoid(function (): void { return; });`},
		{
			Code: `declare function takesVoid(cb: () => void): void;
const cb = () => (0 as any);
takesVoid(cb);`,
			Options: map[string]any{
				"allowReturnAny": true,
			},
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code:   `declare function takesVoid(cb: () => void): void; takesVoid(() => 1);`,
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "nonVoidReturn"}},
		},
		{
			Code:   `declare function takesVoid(cb: () => void): void; takesVoid(async () => {});`,
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "asyncFunc"}},
		},
		{
			Code: `declare function takesVoid(cb: () => void): void;
const cb = () => 1;
takesVoid(cb);`,
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "nonVoidFunc"}},
		},
		{
			Code: `type VoidFn = () => void;
const cb: VoidFn = () => {
  return 1;
};`,
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "nonVoidReturn"}},
		},
	})
}

File: packages/rslint-test-tools/tests/typescript-eslint/rules/strict-void-return.test.ts
import { RuleTester } from '@typescript-eslint/rule-tester';

import { getFixturesRootDir } from '../RuleTester';

const rootDir = getFixturesRootDir();
const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootDir,
    },
  },
});

ruleTester.run('strict-void-return', {
  valid: [
    `
declare function takesVoid(cb: () => void): void;
takesVoid(() => {});
    `,
    `
declare function takesVoid(cb: () => void): void;
takesVoid(function (): void {
  return;
});
    `,
    {
      code: `
declare function takesVoid(cb: () => void): void;
const cb = () => (0 as any);
takesVoid(cb);
      `,
      options: [{ allowReturnAny: true }],
    },
  ],
  invalid: [
    {
      code: `
declare function takesVoid(cb: () => void): void;
takesVoid(() => 1);
      `,
      errors: [{ messageId: 'nonVoidReturn' }],
    },
    {
      code: `
declare function takesVoid(cb: () => void): void;
takesVoid(async () => {});
      `,
      errors: [{ messageId: 'asyncFunc' }],
    },
    {
      code: `
declare function takesVoid(cb: () => void): void;
const cb = () => 1;
takesVoid(cb);
      `,
      errors: [{ messageId: 'nonVoidFunc' }],
    },
    {
      code: `
type VoidFn = () => void;
const cb: VoidFn = () => {
  return 1;
};
      `,
      errors: [{ messageId: 'nonVoidReturn' }],
    },
  ],
});
